freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaHealthDetailsService.java

```diff
@@ -0,0 +1,121 @@
+import java.util.regex.Pattern;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.DetailedStackStatus;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceGroupType;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.health.HealthDetailsFreeIpaResponse;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.health.NodeHealthDetailsFreeIpaResponse;
+import com.sequenceiq.freeipa.client.FreeIpaClient;
+import com.sequenceiq.freeipa.client.FreeIpaClientException;
+import com.sequenceiq.freeipa.client.model.RPCMessage;
+import com.sequenceiq.freeipa.client.model.RPCResponse;
+import com.sequenceiq.freeipa.entity.InstanceGroup;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.service.freeipa.FreeIpaClientFactory;
+
+@Service
+public class FreeIpaHealthDetailsService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaHealthDetailsService.class);
+
+    private static final String EXTERNAL_COMMAND_OUTPUT = "ExternalCommandOutput";
+
+    private static final String STATUS_OK = "OK";
+
+    private static final int STATUS_GROUP = 2;
+
+    private static final String MESSAGE_UNAVAILABLE = "Message Unavailable";
+
+    private static final Pattern RESULT_PATTERN = Pattern.compile("(ecure port|: TCP) \\([0-9]*\\): (.*)");
+
+    private static final Pattern NEW_NODE_PATTERN = Pattern.compile("Check connection from master to remote replica '(.[^\']*)");
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaClientFactory freeIpaClientFactory;
+
+    public HealthDetailsFreeIpaResponse getHealthDetails(String environmentCrn, String accountId) {
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(environmentCrn, accountId);
+        String masterCN = findMasterCN(stack);
+        RPCResponse<Boolean> rpcResponse = null;
+        try {
+            rpcResponse = checkFreeIpaHealth(stack, masterCN);
+        } catch (FreeIpaClientException e) {
+            LOGGER.error("Unable to check the health of FreeIPA.", e);
+        }
+        return createResponse(stack, rpcResponse);
+    }
+
+    private HealthDetailsFreeIpaResponse createResponse(Stack stack, RPCResponse<Boolean> rpcResponse) {
+        HealthDetailsFreeIpaResponse response = new HealthDetailsFreeIpaResponse();
+        response.setEnvironmentCrn(stack.getEnvironmentCrn());
+        response.setCrn(stack.getResourceCrn());
+        if (rpcResponse == null) {
+            response.setStatus(DetailedStackStatus.UNHEALTHY.getStatus());
+        } else {
+            response.setStatus(rpcResponse.getResult() ? DetailedStackStatus.PROVISIONED.getStatus()
+                    : DetailedStackStatus.UNHEALTHY.getStatus());
```

Comment:
this case need unit test also, I couldn't find any assert for `UNHEALTHY` status

Reply:
added